File: app/src/main/java/com/compuspulse/data/repository/SearchRepository.java
package com.compuspulse.data.repository;

import com.compuspulse.data.remote.http.NetworkClient;
import com.compuspulse.data.remote.model.ApiResponse;
import com.compuspulse.data.remote.model.Article;
import com.compuspulse.data.remote.model.PageResponse;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SearchRepository extends BaseRemoteRepository {

    private static final int DEFAULT_PAGE_SIZE = 20;
    private static final Type SEARCH_RESPONSE_TYPE =
            new TypeToken<ApiResponse<PageResponse<Article>>>() { }.getType();

    public void searchArticles(String keyword, RepositoryCallback<List<Article>> callback) {
        Map<String, String> queryParams = new HashMap<>();
        queryParams.put("page_size", String.valueOf(DEFAULT_PAGE_SIZE));

        Map<String, String> formParams = new HashMap<>();
        formParams.put("k", keyword);

        networkClient.postForm("article/query/0/json", queryParams, formParams, SEARCH_RESPONSE_TYPE,
                new NetworkClient.ResponseCallback<PageResponse<Article>>() {
                    @Override
                    public void onSuccess(ApiResponse<PageResponse<Article>> response) {
                        dispatchArticleList(response, callback, "搜索失败");
                    }

                    @Override
                    public void onError(String message) {
                        callback.onError(message);
                    }
                });
    }
}

File: app/src/main/java/com/compuspulse/data/repository/SearchRepository.kt
package com.compuspulse.data.repository

import com.compuspulse.data.remote.model.ApiResponse
import com.compuspulse.data.remote.model.Article
import com.compuspulse.data.remote.model.PageResponse
import com.google.gson.reflect.TypeToken
import java.lang.reflect.Type

class SearchRepository : BaseRemoteRepository() {

    suspend fun searchArticles(keyword: String): Result<List<Article>> = runCatching {
        val queryParams = HashMap<String, String>()
        queryParams["page_size"] = DEFAULT_PAGE_SIZE.toString()

        val formParams = HashMap<String, String>()
        formParams["k"] = keyword

        val response: ApiResponse<PageResponse<Article>> = networkClient.postForm(
            "article/query/0/json",
            queryParams,
            formParams,
            SEARCH_RESPONSE_TYPE
        )
        requireArticleList(response, "搜索失败")
    }

    companion object {
        private const val DEFAULT_PAGE_SIZE = 20
        private val SEARCH_RESPONSE_TYPE: Type =
            object : TypeToken<ApiResponse<PageResponse<Article>>>() {}.type
    }
}
